src/fdc3/dotnet/DesktopAgent/src/DesktopAgent/UserChannel.cs

```diff
@@ -0,0 +1,149 @@
+ * regarding copyright ownership. Unless required by applicable law or agreed
+ * to in writing, software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+using System.Collections.Concurrent;
+using System.Reactive.Linq;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Microsoft.Extensions.Logging;
+using MorganStanley.ComposeUI.Messaging;
+
+namespace DesktopAgent
+{
+    public class UserChannel
+    {
+        public string Name { get; }
+        private readonly ConcurrentDictionary<string, string> _contexts = new ConcurrentDictionary<string, string>();
+        private string? _lastContext = null;
+        private readonly string _topicRoot;
+        private readonly IMessageRouter _messageRouter;
+        private IAsyncObserver<TopicMessage> _broadcastObserver;
+        private ILogger<UserChannel> _logger;
+
+
+        public UserChannel(string name, IMessageRouter messageRouter, ILogger<UserChannel>? logger)
+        {
+            Name = name;
+            _topicRoot = $"composeui/fdc3/userchannels/{name}/";
+            _messageRouter = messageRouter;
+
+            _logger = logger;
+        }
+
+        public async ValueTask Connect()
+        {
+            await _messageRouter.ConnectAsync();
+
+            _broadcastObserver = AsyncObserver.Create<TopicMessage>(x => HandleBroadcast(x.Payload?.ToString()));
+
+            var broadcastSubscribing = _messageRouter.SubscribeAsync(_topicRoot + "broadcast", _broadcastObserver);
+            await broadcastSubscribing;
+
+            LogConnected();
+        }
+
+        internal ValueTask HandleBroadcast(string? payload)
+        {
+            if (string.IsNullOrEmpty(payload))
+            {
+                LogNullOrEmptyBroadcast();
+                return ValueTask.CompletedTask;
+            }
+            LogPayload(payload);
+            JsonNode ctx;
+            try
+            {
+                ctx = JsonNode.Parse(payload, new JsonNodeOptions() { PropertyNameCaseInsensitive = true })!;
+            }
+            catch (JsonException)
+            {
+                LogInvalidPayloadJson();
+                return ValueTask.CompletedTask;
+            }
+            var contextType = (string?) ctx!["type"];
+
+            if (string.IsNullOrEmpty(contextType))
+            {
+                LogMissingContextType();
+                return ValueTask.CompletedTask;
+            }
+            // Although AddOrUpdate is not atomic, MessageRouter will serialize incoming broadcasts
+            // so no race condition can happen due to the non-atomic nature of AddOrUpdate
+
+            _contexts.AddOrUpdate(contextType, _ => UpdateLastContext(payload), (_, _) => UpdateLastContext(payload));
```

Comment:
Do you actually need that method, given this is the only place where it's called from? How about:

```cs
_contexts[contextType] = payload;
_lastContext = payoad;
```

Reply:
Or I'll just use this piece of code instead of overengineering :)